keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+    * [Application Health](#application-health)
+    * [Application Update](#application-update)
+    * [Application List](#application-list)
+    * [Application Watch](#application-watch)
+    * [Application Deletion](#application-deletion)
+ * [Implementation](#implementation)
+ * [API](#api)
+    * [Validation](#validation)
+    * [Generation Incrementation](#generation-incrementation)
+    * [Namespaces](#namespaces)
+    * [Composition (Is A)](#composition-is-a)
+    * [Aggregation (Has A)](#aggregation-has-a)
+    * [Selecting Components](#selecting-components)
+    * [Retrieving Applications](#retrieving-applications)
+    * [Retrieving Dependencies](#retrieving-dependencies)
+    * [Health Checks](#health-checks)
+ * [Application Controller](#application-controller)
+    * [Garbage Collection](#garbage-collection)
+    * [Installation Status Reporting](#installation-status-reporting)
+    * [Health Check Pods](#health-check-pods)
+    * [Health Status Reporting.](#health-status-reporting)
+    * [Generation Observation](#generation-observation)
+    * [Adopting Existing Components](#adopting-existing-components)
+ * [kubectl](#kubectl)
+ * [UI Considerations](#ui-considerations)
+    * [Discovering Application Components and Dependencies](#discovering-application-components-and-dependencies)
+    * [Discovering an Object's Application](#discovering-an-objects-application)
+    * [Application Status](#application-status)
+ * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+ * [Planned](#planned)
+* [Drawbacks](#drawbacks)
+ * [Increased Memory, CPU, and Network Requirements](#increased-memory-cpu-and-network-requirements)
+ * [Increased API Server Storage Requirements](#increased-api-server-storage-requirements)
+* [Alternatives](#alternatives)
+ * [Use Labeling Conventions](#use-labeling-conventions)
+ * [Use Custom Resource Definitions](#use-custom-resource-definitions)
+ * [Application Type Aggregation](#application-type-aggregation)
+ * [Use the Garbage Collector Instead of an Application Controller](#use-the-garbage-collector-instead-of-an-application-controller)
+
+
+
+
+
+
+## Summary
+Kubernetes has many primitives for managing workloads (e.g. Pods, ReplicaSets, Deployments, DaemonSets and 
+StatefulSets), storage (e.g. PersistentVolumeClaims and PersistentVolumes), and networking (e.g. Services, 
+Headless Services, and Ingeresses). When these primitives are aggregated to provide a service to an end user or to 
+another system, the whole becomes something more than the individual parts. Instead of a set of loosely coupled 
+workloads and their corresponding storage and networking, we have an application. 
+
+Currently, there is no standard way for tools to discover the relationship between the components of an 
+application. Many tools use bespoke schemes (generally involving labeling and/or annotations). This has lead to a 
```

Comment:
For comparison, the label convention proposal:
https://docs.google.com/document/d/1EVy0wRJRm5nogkHl38fNKbFrhERmSL_CLNE4cxcsc_M/edit

Openshift example:
https://github.com/openshift/library/blob/master/official/mysql/templates/mysql-persistent.json#L10